packages/medusa/src/strategies/price-selection.ts

```diff
@@ -27,12 +32,112 @@ class PriceSelectionStrategy extends AbstractPriceSelectionStrategy {
     return new PriceSelectionStrategy({
       manager: manager,
       moneyAmountRepository: this.moneyAmountRepository_,
+      featureFlagRouter: this.featureFlagRouter_,
     })
   }
 
   async calculateVariantPrice(
     variant_id: string,
     context: PriceSelectionContext
+  ): Promise<PriceSelectionResult> {
+    if (
+      this.featureFlagRouter_.isFeatureEnabled(
+        TaxInclusivePricingFeatureFlag.key
+      )
+    ) {
+      return this.calculateVariantPrice_new(variant_id, context)
+    }
+    return this.calculateVariantPrice_old(variant_id, context)
+  }
+
+  private async calculateVariantPrice_new(
+    variant_id: string,
+    context: PriceSelectionContext
+  ): Promise<PriceSelectionResult> {
+    const moneyRepo = this.manager_.getCustomRepository(
+      this.moneyAmountRepository_
+    )
+
+    const [prices, count] = await moneyRepo.findManyForVariantInRegion(
+      variant_id,
+      context.region_id,
+      context.currency_code,
+      context.customer_id,
+      context.include_discount_prices,
+      true
+    )
+
+    const result: PriceSelectionResult = {
+      originalPrice: null,
+      calculatedPrice: null,
+      prices,
+      originalPriceIncludesTax: null,
+      calculatedPriceIncludesTax: null,
+    }
+
+    if (!count || !context) {
+      return result
+    }
+
+    const taxRate = context.tax_rates?.reduce(
+      (accRate: number, nextTaxRate: TaxServiceRate) => {
+        return accRate + (nextTaxRate.rate || 0) / 100
+      },
+      0
+    )
+
+    for (const ma of prices) {
+      const isTaxInclusive = !!(
+        ma.region?.includes_tax ||
+        ma.currency?.includes_tax ||
+        ma.price_list?.includes_tax
+      )
```

Comment:
todo: this should be updated to ensure that `isTaxInclusive` is always defined by what **type** of price we are dealing with (i.e., Currency, Region, or Price List) and not by any of region/currency/price_list being tax inclusive.

```js
let isTaxInclusive = ma.currency?.includes_tax || false // default to the currency defining the tax setting

// This is a price list price so we need to use the price list tax setting
if (ma.price_list) {
  isTaxInclusive = ma.price_list.includes_tax
} else if (ma.region) { // this is a region price so we need to use the region tax setting
  isTaxInclusive = ma.region.includes_tax
}
```